src/compiler/checker.ts

```diff
@@ -14400,12 +14432,126 @@ namespace ts {
+                        if (!local.hasReference && local.valueDeclaration && local.valueDeclaration.kind) {
+                            if (local.valueDeclaration.kind === SyntaxKind.Parameter) {
+                                if (local.valueDeclaration.modifiers) {
+                                    if (getCombinedNodeFlags(local.valueDeclaration) & NodeFlags.Private) {
+                                        error(local.valueDeclaration, Diagnostics._0_is_declared_but_never_used, key);
+                                    }
+                                }
+                                else {
+                                    error(local.valueDeclaration, Diagnostics._0_is_declared_but_never_used, key);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        function checkUnusedModulePrivates(node: ModuleDeclaration): void {
+            if (compilerOptions.noUnusedLocals && !isInAmbientContext(node)) {
+                for (const key in node.locals) {
+                    if (hasProperty(node.locals, key)) {
+                        const local = node.locals[key];
+                        if (!local.hasReference && !local.exportSymbol) {
+                            if ((local.valueDeclaration && local.valueDeclaration.kind)) {
+                                error(local.valueDeclaration, Diagnostics._0_is_declared_but_never_used, key);
+                            }
+                            else if (local.declarations &&
+                                (local.declarations[0].kind === SyntaxKind.InterfaceDeclaration || local.declarations[0].kind === SyntaxKind.ModuleDeclaration)) {
+                                error(local.declarations[0], Diagnostics._0_is_declared_but_never_used, key);
+                            }
+                            else if (local.declarations && local.declarations[0].kind === SyntaxKind.ImportEqualsDeclaration) {
+                                error(local.declarations[0].name, Diagnostics._0_is_declared_but_never_used, key);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        function checkUnusedPrivates(node: ClassDeclaration): void {
+            if (compilerOptions.noUnusedLocals && !isInAmbientContext(node)) {
+                if (node.members) {
+                    for (const member of node.members) {
+                        if (member.kind === SyntaxKind.MethodDeclaration || member.kind === SyntaxKind.PropertyDeclaration) {
+                            if (isPrivateClassElement(member) && !member.symbol.hasReference) {
+                                error(member, Diagnostics._0_is_declared_but_never_used, member.symbol.name);
+                            }
+                        }
+                    }
+                }
+
+                if (node.typeParameters) {

```

Comment:
i would extract this to `checkUnusedTypeParamters` and call it for functions and methods as well.